### ISSUE TITLE
Misleading classification performance when falling back to weighted priors with certain class distributions

### ISSUE THREAD

**Issue by JNewmanUEA** [microsoft/LightGBM#7153]
## Description
When LightGBM fails to train it essentially generates a dummy classifier relying solely on class priors to determine classification. When 'balanced' class_weight is used in combination with certain distributions and quantities of class labels, this leads to potentially misleading classification performance for non-trivial classification tasks. In the cross-fold validation example I provide, 100% classification accuracy is obtained for a 3-class problem with 24 folds. This result occurs irrespective of the training data used. This result gives the impression of data leakage or perhaps some kind of magic! Given this result, I don't think this is a safe method of falling back.

## Reproducible example

```python
"""
Demonstration of LightGBM producing 100% classification accuracy when falling
back to weighted priors for certain class distributions. Hold-one-out cross fold
validation experiment, with 24 folds.
Dr Jacob Newman, University of East Anglia, United Kingdom.
jacob.newman@uea.ac.uk
"""
import numpy as np
import lightgbm as lgb

correct = 0
total = 0

# Gives 100% accuracy
participant_id = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23])
participant_diagnosis = np.array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2])

# 0%
# participant_id = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23])
# participant_diagnosis = np.array([0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2])

# 42%
# participant_id = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23])
# participant_diagnosis = np.array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2])

for key in range(0,24):
    test_label = participant_diagnosis[key]
    test_key = participant_id[key]

    print('--------------------------------------------')
    print(f'Subject {test_key} which is label {test_label}')
    
    train_idx = np.where(participant_id != test_key)[0]
    train_keys = participant_id[train_idx]
    
    training_data = np.zeros((23,1))
    training_labels = np.array(participant_diagnosis[train_keys])
    
    print(f'Training labels:\t\t{training_labels}')
   
    clf = lgb.LGBMClassifier(class_weight='balanced', verbose=-1)
    clf.fit(training_data, training_labels)
    predictions = clf.predict(training_data)
    
    print(f'Predicted labels:\t{predictions}')
    
    if predictions[0] == test_label:
        correct+=1
    total+=1
        
print(f'{correct} out of {total} correct')
```
## Environment info

LightGBM was installed using pip:
lightgbm                  4.6.0                    pypi_0    pypi

**Comment by jameslamb** [comment]
Thanks very much for opening this. I had about 75% of an answer written up but got distracted with other things. I do not think it's data leakage but I do think there might be a bug here.

If you don't here from me with all the details in the next few weeks, please `@` me here and remind me to finish the investigation and post results. Sorry for the delay.

**Comment by JNewmanUEA** [comment]
@jameslamb No pressure, but any updates would be most welcome! Thanks.
